text/0000-deprecate-computed-property-modifier.md

```diff
@@ -0,0 +1,123 @@
+- Start Date: 2019-09-13
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Deprecate the computed `.property()` modifier which can be used to add dependent
+keys to computed properties.
+
+# Motivation
+
+Currently, computed properties can use the `.property` modifier to add dependent
+keys to a computed _after_ the computed has been declared:
+
+```js
+foo: computed('strings', {
+  get() {
+    return this.strings.filter(s => s.includes(this.filterText));
+  }
+}).property('filterText')
+```
+
+In most cases, this ability is redundant, since the dependent keys can be moved
+into the original computed declaration and be equivalent:
+
+```js
+foo: computed('strings', 'filterText', {
+  get() {
+    return this.strings.filter(s => s.includes(this.filterText));
+  }
+})
+```
+
+The one exception is in the case of computed _macros_, specifically macros which
+accept a _function_ such as `filter()` and `map()`:
+
+```js
+foo: filter('strings', function(s) {
+  return s.includes(this.filterText);
+}).property('filterText')
+```
+
+The issue stems from the fact that the inner function can access the class
+instance and use dynamic properties from it, and this access is opaque to the
+macro.
+
+This API is confusing since it bears a strong resemblance to the older style
+of computed property declarations, and at first glance appears to be invalid.
+The few edge-case macros where it does legitimately apply can be rewritten to
+accept more dependent keys, making it fully redundant.
+
+# Transition Path
+
+As mentioned above, macros which receive a callback function as an argument are
+the only valid use of `.property()` in current Ember. Currently, there are two
+such macros in Ember core: `map` and `filter`.
```

Comment:
I might be missing something here, but couple comments:

1. It looks like all `ComputedProperty` instances attach their `dependentKey` using `.property`.

    https://github.com/emberjs/ember.js/blob/master/packages/%40ember/-internals/metal/lib/computed.ts#L617

    Does that mean that the class itself needs to be changed to accept dependentKeys as constructor arguments?
2. Secondly, in the case of `arrayMacros`, it looks like only dependentKey actually being attached using `.property()` is the key that's originally passed in (appended with `.[]`). Which to me sounds like the proposal for `additionalDependentKeys` is unnecessary?
    https://github.com/emberjs/ember.js/blob/master/packages/%40ember/object/lib/computed/reduce_computed_macros.js#L56

Reply:
1. This is more of an implementation detail which is why I didn't address it in the RFC, but yes this would have to change. The most direct way forward would be to make the `property()` api private (e.g. `_property`) but we could change the constructor as well.

2. The key difference with the array macros is that they receive arbitrary _functions_ as arguments, and have no way of knowing if those functions introduce additional dependencies. Take the example from the RFC:

   ```js
   foo: filter('strings', function(s) {
     return s.includes(this.filterText);
   }).property('filterText')
   ```

   The `filter` macro has no way to know that it should update whenever `filterText` has changed, it only knows that whenever the `strings` array has updated it should as well. Everything that happens inside the function argument is a black box to it, completely opaque. This can affect _any_ macro which receives arbitrary functions as an argument, it just so happens that the only ones in Ember's core library are `map` and `filter`.

Reply:
Ah ok, that makes sense, thanks for explaining. 

For what it's worth, using `filter().property()` does not seem like a common case to me. Considering that this is is a breaking change already, it's possible that instead of adding the `additionalDependentKeys` API, the recommendation is to just use `Ember.computed` for these cases instead. 

Another alternative would be to pass in any number of strings, where the first is the enumerable, and all following string args are dependentKeys. This would be more closely aligned with the Ember.computed API.

Reply:
It's definitely not a common case, but `ember-decorators` have received reports of users using it this way (that's actually how I found out about this issue). But yes, those are both reasonable alternatives, I'll add them to the alternatives section

Reply:
> Considering that this is is a breaking change already

I do not see any breaking changes proposed here? Existing syntaxes should be completely compatible with a deprecation warning, have we missed something?

Reply:
Removing `computed().property()` would technically be breaking which is what I think @mehulkar was referring too, but that would happen in Ember v4 and there would be a very long overlap period with the deprecation warning.

Reply:
yep, I was referring to removing `.property()`, but you're right there's a different rfc for that I think.

Reply:
I think the different RFC deprecates function prototype extensions, which are actually subtly different:

```js
foo: function() {

}.property('bar'), // calls a method that was added to Function.prototype

bar: computed(function() {

}).property('baz') // calls a method on the ComputedProperty class
```